capsules/src/ctap.rs

```diff
@@ -0,0 +1,310 @@
+//! Provides userspace with access CTAP devices over any transport
+//! layer (USB HID, BLE, NFC). Currently only USB HID is supported.
+//!
+//! Setup
+//! -----
+//!
+//! You need a device that provides the `hil::usb::UsbController` and
+//! `hil::usb_hid::UsbHid` trait.
+//!
+//! ```rust
+//!     let ctap_send_buffer = static_init!([u8; 64], [0; 64]);
+//!     let ctap_recv_buffer = static_init!([u8; 64], [0; 64]);
+//!
+//!     let (ctap, ctap_driver) = components::ctap::CtapComponent::new(
+//!         &earlgrey::usbdev::USB,
+//!         0x1337, // My important company
+//!         0x0DEC, // My device name
+//!         strings,
+//!         board_kernel,
+//!         ctap_send_buffer,
+//!         ctap_recv_buffer,
+//!     )
+//!     .finalize(components::usb_ctap_component_helper!(lowrisc::usbdev::Usb));
+//!
+//!     ctap.enable();
+//!     ctap.attach();
+//! ```
+//!
+
+use core::cell::Cell;
+use core::marker::PhantomData;
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::hil::usb_hid;
+use kernel::{AppId, AppSlice, Callback, Driver, Grant, ReturnCode, Shared};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Ctap as usize;
+
+pub struct App {
+    callback: OptionalCell<Callback>,
+    recv_buf: Option<AppSlice<Shared, u8>>,
+    send_buf: Option<AppSlice<Shared, u8>>,
+    can_receive: Cell<bool>,
+}
+
+impl Default for App {
+    fn default() -> App {
+        App {
+            callback: OptionalCell::empty(),
+            recv_buf: None,
+            send_buf: None,
+            can_receive: Cell::new(false),
+        }
+    }
+}
+
+pub static mut WRITE_BUF: [u8; 64] = [0; 64];
+pub static mut READ_BUF: [u8; 64] = [0; 64];
```

Comment:
Static mut buffers is a pattern to avoid (#1545). We have `static_init` for that.

Reply:
Removed